Add optional configMaps to revision controller

mrogers950 opened the pull request:
This lets us pass optional configMaps to the controller that will not halt progress if they don't exist.
@deads2k @openshift/sig-security 
re: https://github.com/openshift/cluster-kube-apiserver-operator/pull/185

enj commented on pkg/operator/staticpod/controller/revision/deployment_controller.go:
```diff
@@ -180,6 +184,12 @@ func (c RevisionController) createNewRevision(revision int32) error {
+	for _, name := range c.optionalConfigMaps {
+		_, _, err := resourceapply.SyncConfigMap(c.kubeClient.CoreV1(), c.eventRecorder, c.targetNamespace, name, c.targetNamespace, nameFor(name, revision), ownerRefs)
+		if err != nil {
+			return err
```
uuh, how is optional?

mrogers950 replied:
does not return err if obj == nil

enj replied:
Ah sorry I was thinking this was `ApplyConfigMap`.

deads2k commented on pkg/operator/staticpod/controller/revision/deployment_controller.go:
```diff
@@ -52,16 +54,18 @@ type RevisionController struct {
 func NewRevisionController(
 	targetNamespace string,
 	configMaps []string,
+	optionalConfigMaps []string,
```
no.  Make the normal configmaps and secrets become a struct with an `optional` member.

mrogers950 commented:
Updated.

deads2k commented on pkg/operator/staticpod/controllers.go:
```diff
@@ -33,15 +33,34 @@ type staticPodOperatorControllers struct {
 //    account.
 // 5. MonitoringResourceController - this creates the service monitor used by prometheus to scrape metrics.
-func NewControllers(targetNamespaceName, staticPodName string, command, revisionConfigMaps, revisionSecrets []string,
+func NewControllers(targetNamespaceName, staticPodName string, command, revisionConfigMaps, optionalRevisionConfigMaps, revisionSecrets []string,
```
why isn't this the same revision resource?  I expect that ultimately everything that took the list of resources needs to be updated to handle the idea of an optional resource.  The compile breaks will help you.

mrogers950 replied:
fixed this.

mrogers950 commented:
@deads2k updated.

deads2k commented on pkg/operator/staticpod/controller/installer/installer_controller.go:
```diff
@@ -482,15 +483,15 @@ func (c *InstallerController) ensureInstallerPod(nodeName string, operatorSpec *
-	for _, name := range c.configMaps {
-		args = append(args, fmt.Sprintf("--configmaps=%s", name))
+	for _, cm := range c.configMaps {
+		args = append(args, fmt.Sprintf("--configmaps=%s", cm.Name))
```
You're losing information here.  The fact that your signature changed above suggests that the information needs to preserved all the way to the command this is running.  In this case, `installerpod/cmd.go` needs to know that some items can 404 or you'll just permafail the cluster.

deads2k replied:
Also, add a unit test to that command so you can see it work.

deads2k commented on pkg/operator/staticpod/controller/installer/installer_controller.go:
```diff
@@ -478,19 +479,26 @@ func (c *InstallerController) ensureInstallerPod(nodeName string, operatorSpec *
-		args = append(args, fmt.Sprintf("--configmaps=%s", name))
+	for _, cm := range c.configMaps {
+		if !cm.Optional {
+			args = append(args, fmt.Sprintf("--configmaps=%s", cm.Name))
```
These are args to https://github.com/openshift/library-go/blob/master/pkg/operator/staticpod/installerpod/cmd.go#L131 which does not tolerate the idea of optional resources.  You need optional resources, so that information must be plumbed through and those which are *optional* should be tolerated as missing.

mrogers950 replied:
Ah, I see now. Will fix.

mrogers950 commented:
Updated to allow optional configmaps/secrets in the installer pod command, as well as fixing isLatestRevisionCurrent() so it does not always create a new revision when an optional configmap/secret is missing. I added more tests as well.
@deads2k PTAL

deads2k commented on pkg/operator/staticpod/controller/revision/revision_controller.go:
```diff
@@ -129,29 +134,47 @@ func nameFor(name string, revision int32) string {
-		if apierrors.IsNotFound(err) {
+		existing, err := c.configMapGetter.ConfigMaps(c.targetNamespace).Get(nameFor(cm.Name, revision), metav1.GetOptions{})
+		if apierrors.IsNotFound(err) && !cm.Optional {
 			return false, err.Error()
```
not new, but this should probably be `return false, nil`, right?  otherwise I can't easily add anything.

deads2k replied:
> t new, but this should probably be `return false, nil`, right? otherwise I can't easily add anything.

never mind.  reason, not error. ignore me

deads2k commented on pkg/operator/staticpod/controller/revision/revision_controller.go:
```diff
@@ -129,29 +134,47 @@ func nameFor(name string, revision int32) string {
-		if apierrors.IsNotFound(err) {
+		existing, err := c.secretGetter.Secrets(c.targetNamespace).Get(nameFor(s.Name, revision), metav1.GetOptions{})
+		if apierrors.IsNotFound(err) && !s.Optional {
 			return false, err.Error()
```
same comment

deads2k commented on pkg/operator/staticpod/installerpod/cmd.go:
```diff
@@ -19,6 +19,7 @@ import (
 	"k8s.io/client-go/kubernetes"
 
 	"github.com/openshift/library-go/pkg/config/client"
+	"k8s.io/apimachinery/pkg/api/errors"
```
nit order

deads2k commented on pkg/operator/staticpod/installerpod/cmd.go:
```diff
@@ -133,6 +138,15 @@ func (o *InstallOptions) copyContent() error {
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+		if !errors.IsNotFound(err) {
```
don't need this condition.

deads2k commented on pkg/operator/staticpod/installerpod/cmd.go:
```diff
@@ -133,6 +138,15 @@ func (o *InstallOptions) copyContent() error {
 	}
+	for _, currPrefix := range o.OptionalSecretNamePrefixes {
+		val, err := o.KubeClient.CoreV1().Secrets(o.Namespace).Get(o.nameFor(currPrefix), metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
```
```
if errors.IsNotFound{ 
continue
} 
if err != nil{
 return err
}
secrets = append(
```

deads2k commented:
/lgtm

given where we are, at the slightest hint of trouble, this will be reverted.  Sorry.
